common/include/pcl/common/impl/io.hpp

```diff
@@ -136,7 +151,7 @@ copyPointCloud (const pcl::PointCloud<PointInT> &cloud_in,
 
   if (isSamePointType<PointInT, PointOutT> ())
     // Copy the whole memory block
-    memcpy (&cloud_out[0], &cloud_in[0], cloud_in.size () * sizeof (PointInT));
+    copyPointCloudMemcpy (cloud_in, cloud_out);
```

Comment:
This one doesn't need shenanigans. You can use `std::copy` here

Reply:
I'm sorry that if I use `std::copy(&cloud_in[0], &(cloud_in[0])+cloud_in.size(), &cloud_out[0]);` to replace the old `memset`, the error is that, this sentence will be definitely compiled regardless of the completely difference of  `PointInT` and `PointOutT`, which may cause error when copy those elements.

Perhaps I didn't catch what you means well about using `std::copy`. Sorry for that. :cry:

In my opinion, if I extract the `memcpy` into a templated function, when `PointInT` and `PointOutT` are different types(which also means `isSamePointType<PointInT, PointOutT> ()` returns 0 during compile-time), the call of `copyPointCloudMemcpy` after its template instantiation will do nothing, which will handle the previous warnings.

Reply:
> this sentence will be definitely compiled regardless of the completely difference of PointInT and PointOutT

Oh ofc. This isn't `if constexpr`.

In that case, what if we replace the memcpy in the `copyPointCloudMemcpy` function?

Reply:
Yeah, I think it will be better if we replace the C-style memcpy with modern C++ method.:satisfied:

Reply:
Not sure if `std::copy` is really the best option here. We are hoping that `std::copy` uses memcpy internally (for speed reasons), so why not keep using memcpy directly? Also: how about moving the for-loop below into the currently empty `copyPointCloudMemcpy` and remove this if-else completely?

Reply:
Thanks for your reviews! :blush:

> why not keep using memcpy directly?

I think `std::copy` is more flexible than `memcpy` for no performance loss, and is safer and increases in speed especially for those non-POD types like `vector<...> points;` in `pcl::PointCloud`. I've investigated it and found this page really helpful: https://stackoverflow.com/questions/4707012/is-it-better-to-use-stdmemcpy-or-stdcopy-in-terms-to-performance

> how about moving the for-loop below into the currently empty copyPointCloudMemcpy and remove this if-else completely?

Excellent idea! :grinning: And make the code more readable and prettier. Thank you! I'll fix it then.

Reply:
> We are hoping that std::copy uses memcpy internally (for speed reasons), so why not keep using memcpy directly?
> observations from SO thread and `std::copy` being more flexible/faster than `memcpy`

* `std::copy` can handle types which might have non-trivial copy ctors (PCL supports custom points. Using `copy` allows us to support wider cadre of structs as points, not a big priority to-be-honest)
* depending on the distribution and C stdlib, `memcpy` might not be inlined (since it's a pure C function), while `std::copy` has a higher chance of being inlined (if they don't use `memcpy` themselves, due to being templated)
* Some hardened distributions don't use sse, avx in their C stdlib, so even if `std::copy` is just a `for(auto& a: b){*a = *c; a++;c++;}`, the application might be faster with `copy` than with `memcpy` (if the optimizations are turned on) (this is rare, so essentially, I just have 1 point 😆 )

[Link to copy in libcxx](https://github.com/llvm/llvm-project/blob/main/libcxx/include/algorithm#L1701) (clang's stdlib for c++)
[GCC's implementation](https://github.com/gcc-mirror/gcc/blob/master/libstdc%2B%2B-v3/include/bits/stl_algobase.h#L338) is similar but way more complicated (relying on combination of memmove, memcpy, simple loop, etc.)